File: smilemenu/backend/providers.py
import subprocess
from pathlib import Path
from .item import LauncherItem


def load_stdin(display_columns=None):
    import sys
    items = []

    for line in sys.stdin:
        line = line.strip()
        if not line:
            continue
        items.append(LauncherItem(name=get_display_name(line, display_columns), command=line))

    return items


def load_provider(provider, fields=None):
    items = []

    try:
        result = subprocess.run([provider, "list"], capture_output=True, text=True)
    except Exception as e:
        print(f"Provider failed: {provider}")
        print(e)
        return items

    if result.returncode != 0:
        print(f"Provider error: {provider}")
        if result.stderr:
            print(result.stderr.strip())
        return items

    for line in result.stdout.splitlines():
        line = line.strip()
        if not line:
            continue
        items.append(provider_item(line, fields))

    return items


def provider_item(value, fields=None):
    path = Path(value)
    data = {
        "name": path.name,
        "command": value,
        "icon": "",
        "description": ""
    }
    if path.exists():
        data["icon"] = str(path)

    columns = value.split("\t") if "\t" in value else [value]

    if fields:
        for field in fields:
            try:
                key, mode = field.split(":", 1)
            except ValueError:
                continue
            if mode == "path":
                if key == "name":
                    data[key] = path.name
                elif key == "icon":
                    data[key] = str(path)
            elif mode == "text":
                data[key] = value
            elif mode == "none":
                data[key] = ""
            else:
                try:
                    col_index = int(mode) - 1
                    if 0 <= col_index < len(columns):
                        data[key] = columns[col_index]
                except ValueError:
                    pass

    return LauncherItem(
        name=data["name"],
        command=data["command"],
        icon=data["icon"],
        description=data["description"]
    )

def get_display_name(value, display_columns=None):
    if display_columns is None:
        return value

    columns = value.split("\t")

    selected = []

    for column in display_columns.split(","):
        try:
            index = int(column.strip()) - 1
            if 0 <= index < len(columns):
                selected.append(columns[index])
        except ValueError:
            continue

    return "\t".join(selected)

File: src/providers.cpp
#include "providers.h"
#include <QFileInfo>
#include <QDebug>

AppItem* Providers::fromLine(const QString &line, const QStringList &fields)
{
    QStringList columns = line.split('\t');
    QString value = line;

    QFileInfo info(value);
    QString name = info.fileName();
    QString icon = info.exists() ? value : "";
    QString desc;

    if (!fields.isEmpty()) {
        for (const QString &field : fields) {
            QStringList parts = field.split(':');
            if (parts.size() != 2) continue;
            QString key = parts[0];
            QString mode = parts[1];
            if (mode == "path") {
                if (key == "name")
                    name = info.fileName();
                else if (key == "icon")
                    icon = value;
            } else if (mode == "text") {
                if (key == "name")
                    name = value;
                else if (key == "icon")
                    icon = value;
            } else if (mode == "none") {
                if (key == "name")
                    name = "";
                else if (key == "icon")
                    icon = "";
            } else {
                bool ok;
                int idx = mode.toInt(&ok) - 1;
                if (ok && idx >= 0 && idx < columns.size()) {
                    if (key == "name")
                        name = columns[idx];
                    else if (key == "icon")
                        icon = columns[idx];
                    else if (key == "description")
                        desc = columns[idx];
                }
            }
        }
    }

    if (fields.isEmpty()) {
        name = columns.isEmpty() ? value : columns[0];
    }

    return new AppItem(name, value, icon, desc);
}

QString Providers::getDisplayName(const QString &value, const QString &displayColumns)
{
    if (displayColumns.isEmpty())
        return value;
    QStringList columns = value.split('\t');
    QStringList result;
    QStringList indices = displayColumns.split(',');
    for (const QString &idxStr : indices) {
        bool ok;
        int idx = idxStr.trimmed().toInt(&ok) - 1;
        if (ok && idx >= 0 && idx < columns.size())
            result << columns[idx];
    }
    return result.join('\t');
}
